cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperLeaderFinder.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.strimzi.api.kafka.model.KafkaResources;
+import io.strimzi.certs.CertAndKey;
+import io.strimzi.operator.cluster.ClusterOperator;
+import io.strimzi.operator.cluster.model.Ca;
+import io.strimzi.operator.common.BackOff;
+import io.strimzi.operator.common.operator.resource.SecretOperator;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.net.NetClientOptions;
+import io.vertx.core.net.NetSocket;
+import io.vertx.core.net.PemKeyCertOptions;
+import io.vertx.core.net.PemTrustOptions;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.cert.CertificateFactory;
+import java.util.Base64;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Helper class for finding the leader of a ZK cluster
+ */
+public class ZookeeperLeaderFinder {
+
+    private static final Logger log = LogManager.getLogger(ZookeeperLeaderFinder.class);
+
+    static final int UNKNOWN_LEADER = -1;
+
+    private final Vertx vertx;
+    private final SecretOperator secretOperator;
+    private final Supplier<BackOff> backOffSupplier;
+
+    public ZookeeperLeaderFinder(Vertx vertx, SecretOperator secretOperator, Supplier<BackOff> backOffSupplier) {
+        this.vertx = vertx;
+        this.secretOperator = secretOperator;
+        this.backOffSupplier = backOffSupplier;
+    }
+
+    /*test*/ NetClientOptions clientOptions(CertAndKey coCertKey, Secret clusterCaCertificateSecret)
+            throws GeneralSecurityException, IOException {
+
+        CertificateFactory x509 = CertificateFactory.getInstance("X.509");
+        Base64.Decoder decoder = Base64.getDecoder();
+
+        // Validate the certificates are not corrupt
+        //x509.generateCertificate(new ByteArrayInputStream(decoder.decode(clusterCaCertificateSecret.getData().get(Ca.CA_CRT))));
```

Comment:
I was originally using and testing this in the unit tests, but it complicated the `KafkaAssemblyOperatorMockTest` and others, which don't use a real certificate. We could try to arrange that `KafkaAssemblyOperatorMockTest` doesn't do the validation, if we think it's worth keeping. For example by factoring out a `validateCertificate()` method which we override to a no-op in the test. But there's little to trying to validate the certificates here apart from being able to give more specific error message (i.e. one we control, rather than an exception coming from vertx). Which do you prefer?

Reply:
I've refactored this so we do the validation, but skip it for those tests which use fake certs.